osh/builtin_meta.py

```diff
@@ -366,39 +365,21 @@ def __init__(self, expr_ev):
 
     def Run(self, cmd_val):
         # type: (cmd_value.Argv) -> int
-        blame = cmd_val.arg_locs[0]
-
-        argc = len(cmd_val.argv)
-        if argc != 1:
-            e_usage('Expected 0 command arguments but received %d' % (argc - 1), blame)
-
-        if not cmd_val.typed_args:
-            e_usage('Expected an error argument but received no typed args', blame)
-
-        pos_args = cmd_val.typed_args.pos_args
-        if len(pos_args) != 1:
-            e_usage('Expected 1 positional argument but received %d' % len(pos_args), blame)
-
-        message_val = self.expr_ev.EvalExpr(pos_args[0], blame)
-        if message_val.tag() != value_e.Str:
-            raise error.TypeErr(message_val, 'expected a Str', blame)
-        message = cast(value.Str, message_val).s
-
-        status = 1
-        for arg in cmd_val.typed_args.named_args:
-            name = lexer.TokenVal(arg.name)
-            if name == "status":
-                status_val = self.expr_ev.EvalExpr(arg.value, arg.name)
-                if status_val.tag() != value_e.Int:
-                    raise error.TypeErr(status_val, 'expected an Int', arg.name)
-                status = cast(value.Int, status_val).i
-
-                if status == 0:
-                    e_die("Must be a non-zero integer", arg.name)
-            else:
-                e_usage('Unexpected named argument %r' % name, arg.name)
+        t = typed_args.ReaderFromArgv(cmd_val.typed_args, self.expr_ev)
+
+        message = t.PosStr()
+        status = t.NamedInt("status", 1)
+        t.Done()
+
+        if status == 0:
+            e_die("Status must be a non-zero integer", cmd_val.arg_locs[0])
+
+        if len(cmd_val.argv) > 1:
+            raise error.TypeErrVerbose(
+                'Expected 0 untyped arguments, but got %d' %
```

Comment:
I think eventually we will have a variant of `Reader` for  procs that also checks this, but this is good for now